File: backend/application/services/portfolio.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Literal

from strategies_v2.utils import InputPortfolioDataPoint, PortfolioPosition

TradeAction = Literal["buy", "sell", "sell_short", "buy_to_cover", "invalid"]


@dataclass
class Trade:
    unixtime: int
    ticker: str
    direction: str
    action: TradeAction
    price: float
    qty: float
    deposit_ratio: float
    reason: str = ""
    valid: bool = True

    @property
    def label(self) -> str:
        return {
            "buy": "BUY",
            "sell": "SELL",
            "sell_short": "SELL SHORT",
            "buy_to_cover": "BUY TO COVER",
            "invalid": "INVALID",
        }[self.action]


@dataclass
class Position:
    qty: float
    avg_entry_price: float


@dataclass
class Portfolio:
    initial_deposit: float
    ticker: str
    cash: float = field(init=False)
    positions: dict[str, Position] = field(default_factory=dict)
    realized_pnl: float = field(init=False, default=0.0)
    trades: list[Trade] = field(default_factory=list)
    equity_points: list[tuple[int, float]] = field(default_factory=list)
    last_marks: dict[str, float] = field(default_factory=dict)

    def __post_init__(self) -> None:
        if self.initial_deposit <= 0:
            raise ValueError("initial_deposit must be positive")
        self.cash = float(self.initial_deposit)
        self.ticker = str(self.ticker).strip()

    @property
    def position_qty(self) -> float:
        pos = self.positions.get(self.ticker)
        return 0.0 if pos is None else pos.qty

    @property
    def avg_entry_price(self) -> float | None:
        pos = self.positions.get(self.ticker)
        return None if pos is None else pos.avg_entry_price

    def equity(self, mark_price: float | dict[str, float]) -> float:
        if isinstance(mark_price, dict):
            marks = {str(k).strip(): float(v) for k, v in mark_price.items()}
        else:
            marks = {self.ticker: float(mark_price)}
        for t, px in marks.items():
            if px > 0:
                self.last_marks[t] = px
        total = self.cash
        for t, pos in self.positions.items():
            px = marks.get(t, self.last_marks.get(t, pos.avg_entry_price))
            total += pos.qty * px
        return total

    def apply_market_order(
        self,
        *,
        ticker: str | None = None,
        direction: str,
        deposit_ratio: float,
        price: float,
        unixtime: int,
        reason: str = "",
        cash_basis: float | None = None,
    ) -> None:
        t = str(ticker or self.ticker).strip()
        if not t:
            self._record_invalid_order(
                ticker=self.ticker,
                direction=direction,
                deposit_ratio=deposit_ratio,
                price=price,
                unixtime=unixtime,
                reason="ticker is required",
            )
            return
        d = direction.lower().strip()
        dr = float(deposit_ratio)
        if dr <= 0 or dr > 1:
            self._record_invalid_order(
                ticker=t,
                direction=d,
                deposit_ratio=dr,
                price=price,
                unixtime=unixtime,
                reason="deposit_ratio must be in (0, 1]",
            )
            return
        if price <= 0:
            self._record_invalid_order(
                ticker=t,
                direction=d,
                deposit_ratio=dr,
                price=price,
                unixtime=unixtime,
                reason="price must be positive",
            )
            return
        if d == "buy":
            pos = self.positions.get(t)
            if pos is not None and pos.qty < 0:
                qty = abs(pos.qty) * dr
                spend = qty * price
                if spend > self.cash + 1e-9:
                    self._record_invalid_order(
                        ticker=t,
                        direction=d,
                        deposit_ratio=dr,
                        price=price,
                        unixtime=unixtime,
                        reason="insufficient cash for market_order batch",
                    )
                    return
                pnl_leg = qty * (pos.avg_entry_price - price)
                self.realized_pnl += pnl_leg
                self.cash -= spend
                pos.qty += qty
                if abs(pos.qty) <= 1e-12:
                    self.positions.pop(t, None)
                self.trades.append(
                    Trade(
                        unixtime=unixtime,
                        ticker=t,
                        direction="buy",
                        action="buy_to_cover",
                        price=price,
                        qty=qty,
                        deposit_ratio=dr,
                        reason=reason,
                    )
                )
                return

            basis = self.cash if cash_basis is None else float(cash_basis)
            spend = basis * dr
            if spend <= 0:
                return
            if spend > self.cash + 1e-9:
                self._record_invalid_order(
                    ticker=t,
                    direction=d,
                    deposit_ratio=dr,
                    price=price,
                    unixtime=unixtime,
                    reason="insufficient cash for market_order batch",
                )
                return
            qty = spend / price
            if pos is None or pos.qty <= 0:
                self.positions[t] = Position(qty=qty, avg_entry_price=price)
            else:
                total_qty = pos.qty + qty
                pos.avg_entry_price = (
                    pos.avg_entry_price * pos.qty + price * qty
                ) / total_qty
                pos.qty = total_qty
            self.cash -= spend
            self.trades.append(
                Trade(
                    unixtime=unixtime,
                    ticker=t,
                    direction="buy",
                    action="buy",
                    price=price,
                    qty=qty,
                    deposit_ratio=dr,
                    reason=reason,
                )
            )
        elif d == "sell":
            pos = self.positions.get(t)
            if pos is not None and pos.qty > 0:
                qty = pos.qty * dr
                proceeds = qty * price
                pnl_leg = qty * (price - pos.avg_entry_price)
                self.realized_pnl += pnl_leg
                self.cash += proceeds
                pos.qty -= qty
                if pos.qty <= 1e-12:
                    self.positions.pop(t, None)
                self.trades.append(
                    Trade(
                        unixtime=unixtime,
                        ticker=t,
                        direction="sell",
                        action="sell",
                        price=price,
                        qty=qty,
                        deposit_ratio=dr,
                        reason=reason,
                    )
                )
                return

            basis = self.equity(self.last_marks) if self.last_marks else self.initial_deposit
            if basis <= 0:
                basis = self.initial_deposit
            qty = basis * dr / price
            proceeds = qty * price
            self.cash += proceeds
            if pos is None or pos.qty >= 0:
                self.positions[t] = Position(qty=-qty, avg_entry_price=price)
            else:
                open_qty = abs(pos.qty)
                total_qty = open_qty + qty
                pos.avg_entry_price = (pos.avg_entry_price * open_qty + price * qty) / total_qty
                pos.qty -= qty
            self.trades.append(
                Trade(
                    unixtime=unixtime,
                    ticker=t,
                    direction="sell",
                    action="sell_short",
                    price=price,
                    qty=qty,
                    deposit_ratio=dr,
                    reason=reason,
                )
            )
        else:
            self._record_invalid_order(
                ticker=t,
                direction=d,
                deposit_ratio=dr,
                price=price,
                unixtime=unixtime,
                reason=f"Unsupported direction: {direction!r}",
            )

    def apply_market_orders(
        self,
        orders: list,
        *,
        prices: dict[str, float],
        unixtime: int,
        reason: str = "",
    ) -> None:
        sell_orders = []
        buy_orders = []
        for item in orders:
            d = str(item.direction).lower().strip()
            if d == "sell":
                sell_orders.append(item)
            elif d == "buy":
                buy_orders.append(item)
            else:
                self._record_invalid_order_from_item(
                    item,
                    price=0.0,
                    unixtime=unixtime,
                    reason=f"Unsupported direction: {item.direction!r}",
                )

        for item in sell_orders:
            t = str(item.ticker).strip()
            px = prices.get(t)
            if px is None:
                self._record_invalid_order_from_item(
                    item,
                    price=0.0,
                    unixtime=unixtime,
                    reason=f"no fill price available for ticker {t!r}",
                )
                continue
            self.apply_market_order(
                ticker=t,
                direction=item.direction,
                deposit_ratio=item.deposit_ratio,
                price=px,
                unixtime=unixtime,
                reason=reason,
            )

        batch_cash = self.cash
        valid_buy_orders = []
        total_spend = 0.0
        for item in buy_orders:
            t = str(item.ticker).strip()
            px = prices.get(t)
            dr = float(item.deposit_ratio)
            if px is None:
                self._record_invalid_order_from_item(
                    item,
                    price=0.0,
                    unixtime=unixtime,
                    reason=f"no fill price available for ticker {t!r}",
                )
                continue
            if dr <= 0 or dr > 1:
                self._record_invalid_order_from_item(
                    item,
                    price=px,
                    unixtime=unixtime,
                    reason="deposit_ratio must be in (0, 1]",
                )
                continue
            valid_buy_orders.append((item, px))
            total_spend += dr * batch_cash
        if total_spend > self.cash + 1e-9:
            for item, px in valid_buy_orders:
                self._record_invalid_order_from_item(
                    item,
                    price=px,
                    unixtime=unixtime,
                    reason="market_order buy batch exceeds available cash",
                )
            return
        for item, px in valid_buy_orders:
            t = str(item.ticker).strip()
            self.apply_market_order(
                ticker=t,
                direction=item.direction,
                deposit_ratio=item.deposit_ratio,
                price=px,
                unixtime=unixtime,
                reason=reason,
                cash_basis=batch_cash,
            )

    def _record_invalid_order_from_item(
        self,
        item: Any,
        *,
        price: float,
        unixtime: int,
        reason: str,
    ) -> None:
        self._record_invalid_order(
            ticker=getattr(item, "ticker", self.ticker),
            direction=str(getattr(item, "direction", "")),
            deposit_ratio=float(getattr(item, "deposit_ratio", 0.0)),
            price=price,
            unixtime=unixtime,
            reason=reason,
        )

    def _record_invalid_order(
        self,
        *,
        ticker: str,
        direction: str,
        deposit_ratio: float,
        price: float,
        unixtime: int,
        reason: str,
    ) -> None:
        t = str(ticker or self.ticker).strip()
        self.trades.append(
            Trade(
                unixtime=unixtime,
                ticker=t or self.ticker,
                direction=str(direction).lower().strip(),
                action="invalid",
                price=float(price) if price > 0 else 0.0,
                qty=0.0,
                deposit_ratio=float(deposit_ratio),
                reason=reason,
                valid=False,
            )
        )

    def record_equity(self, unixtime: int, mark_price: float | dict[str, float]) -> None:
        self.equity_points.append((unixtime, self.equity(mark_price)))

    def to_portfolio_datapoint(self) -> InputPortfolioDataPoint:
        positions: list[PortfolioPosition] = []
        eq = self.equity(self.last_marks) if self.last_marks else self.initial_deposit
        denom = eq if eq > 0 else self.initial_deposit
        for t in sorted(self.positions):
            pos = self.positions[t]
            px = self.last_marks.get(t, pos.avg_entry_price)
            ratio = (abs(pos.qty) * px / denom) if denom > 0 else 0.0
            positions.append(
                PortfolioPosition(
                    ticker=t,
                    order_type="long" if pos.qty >= 0 else "short",
                    deposit_ratio=max(0.0, min(1.0, float(ratio))),
                    volume_weighted_avg_entry_price=float(pos.avg_entry_price),
                )
            )
        return InputPortfolioDataPoint(kind="portfolio", positions=positions)

File: backend/strategies_v2/portfolio.rs
use crate::utils::{InputDataPoint, OutputDataPoint, PortfolioPosition};
use serde::Serialize;
use std::collections::{BTreeMap, BTreeSet};

const EPSILON: f64 = 1e-9;

#[derive(Debug, Clone)]
pub struct Position {
    pub qty: f64,
    pub avg_entry_price: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct Trade {
    pub unixtime: i64,
    pub ticker: String,
    pub direction: String,
    pub action: String,
    pub price: f64,
    pub qty: f64,
    pub deposit_ratio: f64,
    pub position_before_order: f64,
    pub position_after_order_filled: f64,
    pub reason: String,
    pub valid: bool,
}

impl Trade {
    pub fn label(&self) -> &'static str {
        match self.action.as_str() {
            "buy" => "BUY",
            "sell" => "SELL",
            "sell_short" => "SELL SHORT",
            "buy_to_cover" => "BUY TO COVER",
            _ => "INVALID",
        }
    }
}

pub struct Portfolio {
    initial_deposit: f64,
    primary_ticker: String,
    max_leverage: f64,
    pub cash: f64,
    pub positions: BTreeMap<String, Position>,
    pub trades: Vec<Trade>,
    pub last_marks: BTreeMap<String, f64>,
}

impl Portfolio {
    pub fn new(initial_deposit: f64, primary_ticker: String, max_leverage: f64) -> Self {
        Self {
            initial_deposit,
            primary_ticker,
            max_leverage,
            cash: initial_deposit,
            positions: BTreeMap::new(),
            trades: Vec::new(),
            last_marks: BTreeMap::new(),
        }
    }

    fn qty(&self, ticker: &str) -> f64 {
        self.positions
            .get(ticker)
            .map_or(0.0, |position| position.qty)
    }

    pub fn equity(&mut self, marks: &BTreeMap<String, f64>) -> f64 {
        for (ticker, price) in marks {
            if *price > 0.0 {
                self.last_marks.insert(ticker.clone(), *price);
            }
        }
        self.cash
            + self
                .positions
                .iter()
                .map(|(ticker, position)| {
                    let price = marks
                        .get(ticker)
                        .or_else(|| self.last_marks.get(ticker))
                        .copied()
                        .unwrap_or(position.avg_entry_price);
                    position.qty * price
                })
                .sum::<f64>()
    }

    fn gross_exposure(
        &self,
        marks: &BTreeMap<String, f64>,
        quantities: &BTreeMap<String, f64>,
    ) -> f64 {
        quantities
            .iter()
            .map(|(ticker, qty)| {
                let fallback = self
                    .positions
                    .get(ticker)
                    .map_or(0.0, |position| position.avg_entry_price);
                let price = marks
                    .get(ticker)
                    .or_else(|| self.last_marks.get(ticker))
                    .copied()
                    .unwrap_or(fallback);
                qty.abs() * price.max(0.0)
            })
            .sum()
    }

    fn exceeds_leverage(&mut self, ticker: &str, projected_qty: f64, price: f64) -> bool {
        let mut mark = BTreeMap::new();
        mark.insert(ticker.to_owned(), price);
        let equity = self.equity(&mark);
        if equity <= 0.0 {
            return true;
        }
        let mut quantities: BTreeMap<String, f64> = self
            .positions
            .iter()
            .map(|(name, position)| (name.clone(), position.qty))
            .collect();
        if projected_qty.abs() <= 1e-12 {
            quantities.remove(ticker);
        } else {
            quantities.insert(ticker.to_owned(), projected_qty);
        }
        self.gross_exposure(&mark, &quantities) > equity * self.max_leverage + EPSILON
    }

    fn invalid(
        &mut self,
        ticker: &str,
        direction: &str,
        deposit_ratio: f64,
        price: f64,
        qty: f64,
        unixtime: i64,
        explanation: &str,
        reason: &str,
    ) {
        let detail = if explanation.trim().is_empty() {
            reason.to_owned()
        } else {
            format!("{}: {}", explanation.trim(), reason)
        };
        let before = self.qty(ticker);
        self.trades.push(Trade {
            unixtime,
            ticker: ticker.to_owned(),
            direction: direction.trim().to_lowercase(),
            action: "invalid".to_owned(),
            price: price.max(0.0),
            qty: qty.max(0.0),
            deposit_ratio,
            position_before_order: before,
            position_after_order_filled: before,
            reason: detail,
            valid: false,
        });
    }

    #[allow(clippy::too_many_arguments)]
    fn apply_one(
        &mut self,
        ticker: &str,
        direction: &str,
        deposit_ratio: f64,
        price: f64,
        unixtime: i64,
        explanation: &str,
        cash_basis: Option<f64>,
    ) {
        if ticker.trim().is_empty() {
            let primary = self.primary_ticker.clone();
            self.invalid(
                &primary,
                direction,
                deposit_ratio,
                price,
                0.0,
                unixtime,
                explanation,
                "ticker is required",
            );
            return;
        }
        if !(deposit_ratio > 0.0 && deposit_ratio <= 1.0) {
            self.invalid(
                ticker,
                direction,
                deposit_ratio,
                price,
                0.0,
                unixtime,
                explanation,
                "deposit_ratio must be in (0, 1]",
            );
            return;
        }
        if price <= 0.0 {
            self.invalid(
                ticker,
                direction,
                deposit_ratio,
                price,
                0.0,
                unixtime,
                explanation,
                "price must be positive",
            );
            return;
        }
        let direction = direction.trim().to_lowercase();
        let before = self.qty(ticker);
        if direction == "buy" {
            if before < 0.0 {
                let qty = before.abs() * deposit_ratio;
                let spend = qty * price;
                if spend > self.cash + EPSILON {
                    self.invalid(
                        ticker,
                        &direction,
                        deposit_ratio,
                        price,
                        qty,
                        unixtime,
                        explanation,
                        "insufficient cash for market_order batch",
                    );
                    return;
                }
                self.cash -= spend;
                let after = before + qty;
                if after.abs() <= 1e-12 {
                    self.positions.remove(ticker);
                } else if let Some(position) = self.positions.get_mut(ticker) {
                    position.qty = after;
                }
                self.push_trade(
                    ticker,
                    &direction,
                    "buy_to_cover",
                    price,
                    qty,
                    deposit_ratio,
                    before,
                    if after.abs() <= 1e-12 { 0.0 } else { after },
                    unixtime,
                    explanation,
                );
                return;
            }
            let basis = cash_basis.unwrap_or(self.cash);
            let spend = basis * deposit_ratio;
            if spend <= 0.0 {
                return;
            }
            let qty = spend / price;
            if spend > self.cash + EPSILON {
                self.invalid(
                    ticker,
                    &direction,
                    deposit_ratio,
                    price,
                    qty,
                    unixtime,
                    explanation,
                    "insufficient cash for market_order batch",
                );
                return;
            }
            if self.exceeds_leverage(ticker, before + qty, price) {
                self.invalid(
                    ticker,
                    &direction,
                    deposit_ratio,
                    price,
                    qty,
                    unixtime,
                    explanation,
                    "max_leverage exceeded",
                );
                return;
            }
            self.cash -= spend;
            if let Some(position) = self.positions.get_mut(ticker) {
                let total = position.qty + qty;
                position.avg_entry_price =
                    (position.avg_entry_price * position.qty + price * qty) / total;
                position.qty = total;
            } else {
                self.positions.insert(
                    ticker.to_owned(),
                    Position {
                        qty,
                        avg_entry_price: price,
                    },
                );
            }
            self.push_trade(
                ticker,
                &direction,
                "buy",
                price,
                qty,
                deposit_ratio,
                before,
                before + qty,
                unixtime,
                explanation,
            );
        } else if direction == "sell" {
            if before > 0.0 {
                let qty = before * deposit_ratio;
                self.cash += qty * price;
                let after = before - qty;
                if after <= 1e-12 {
                    self.positions.remove(ticker);
                } else if let Some(position) = self.positions.get_mut(ticker) {
                    position.qty = after;
                }
                self.push_trade(
                    ticker,
                    &direction,
                    "sell",
                    price,
                    qty,
                    deposit_ratio,
                    before,
                    if after <= 1e-12 { 0.0 } else { after },
                    unixtime,
                    explanation,
                );
                return;
            }
            let mut basis = if self.last_marks.is_empty() {
                self.initial_deposit
            } else {
                let marks = self.last_marks.clone();
                self.equity(&marks)
            };
            if basis <= 0.0 {
                basis = self.initial_deposit;
            }
            let qty = basis * deposit_ratio / price;
            if self.exceeds_leverage(ticker, before - qty, price) {
                self.invalid(
                    ticker,
                    &direction,
                    deposit_ratio,
                    price,
                    qty,
                    unixtime,
                    explanation,
                    "max_leverage exceeded",
                );
                return;
            }
            self.cash += qty * price;
            if let Some(position) = self.positions.get_mut(ticker) {
                let open_qty = position.qty.abs();
                let total = open_qty + qty;
                position.avg_entry_price =
                    (position.avg_entry_price * open_qty + price * qty) / total;
                position.qty -= qty;
            } else {
                self.positions.insert(
                    ticker.to_owned(),
                    Position {
                        qty: -qty,
                        avg_entry_price: price,
                    },
                );
            }
            self.push_trade(
                ticker,
                &direction,
                "sell_short",
                price,
                qty,
                deposit_ratio,
                before,
                before - qty,
                unixtime,
                explanation,
            );
        } else {
            self.invalid(
                ticker,
                &direction,
                deposit_ratio,
                price,
                0.0,
                unixtime,
                explanation,
                &format!("Unsupported direction: {direction:?}"),
            );
        }
    }

    #[allow(clippy::too_many_arguments)]
    fn push_trade(
        &mut self,
        ticker: &str,
        direction: &str,
        action: &str,
        price: f64,
        qty: f64,
        deposit_ratio: f64,
        before: f64,
        after: f64,
        unixtime: i64,
        reason: &str,
    ) {
        self.trades.push(Trade {
            unixtime,
            ticker: ticker.to_owned(),
            direction: direction.to_owned(),
            action: action.to_owned(),
            price,
            qty,
            deposit_ratio,
            position_before_order: before,
            position_after_order_filled: after,
            reason: reason.to_owned(),
            valid: true,
        });
    }

    pub fn apply_outputs(
        &mut self,
        outputs: &[OutputDataPoint],
        prices: &BTreeMap<String, f64>,
        unixtime: i64,
        single_ticker: bool,
    ) {
        let mut pending: Vec<(&str, &str, f64, &str, f64)> = Vec::new();
        for output in outputs {
            let OutputDataPoint::MarketOrder {
                ticker,
                direction,
                deposit_ratio,
                short_explanation,
            } = output
            else {
                continue;
            };
            let price = prices.get(ticker).copied().or_else(|| {
                if single_ticker {
                    prices.values().next().copied()
                } else {
                    None
                }
            });
            if direction.trim().eq_ignore_ascii_case("buy") && self.qty(ticker) >= 0.0 {
                if let Some(price) = price {
                    pending.push((ticker, direction, *deposit_ratio, short_explanation, price));
                    continue;
                }
            }
            self.flush_buys(&mut pending, unixtime);
            match price {
                Some(price) => self.apply_one(
                    ticker,
                    direction,
                    *deposit_ratio,
                    price,
                    unixtime,
                    short_explanation,
                    None,
                ),
                None => self.invalid(
                    ticker,
                    direction,
                    *deposit_ratio,
                    0.0,
                    0.0,
                    unixtime,
                    short_explanation,
                    &format!("no fill price available for ticker {ticker:?}"),
                ),
            }
        }
        self.flush_buys(&mut pending, unixtime);
    }

    fn flush_buys(&mut self, pending: &mut Vec<(&str, &str, f64, &str, f64)>, unixtime: i64) {
        if pending.is_empty() {
            return;
        }
        let batch_cash = self.cash;
        if pending
            .iter()
            .map(|(_, _, ratio, _, _)| ratio * batch_cash)
            .sum::<f64>()
            > self.cash + EPSILON
        {
            for (ticker, direction, ratio, explanation, price) in pending.drain(..) {
                self.invalid(
                    ticker,
                    direction,
                    ratio,
                    price,
                    ratio * batch_cash / price,
                    unixtime,
                    explanation,
                    "market_order buy batch exceeds available cash",
                );
            }
        } else {
            for (ticker, direction, ratio, explanation, price) in pending.drain(..) {
                self.apply_one(
                    ticker,
                    direction,
                    ratio,
                    price,
                    unixtime,
                    explanation,
                    Some(batch_cash),
                );
            }
        }
    }

    pub fn input_point(&mut self) -> InputDataPoint {
        let marks = self.last_marks.clone();
        let equity = if marks.is_empty() {
            self.initial_deposit
        } else {
            self.equity(&marks)
        };
        let denominator = if equity > 0.0 {
            equity
        } else {
            self.initial_deposit
        };
        let positions = self
            .positions
            .iter()
            .map(|(ticker, position)| {
                let price = self
                    .last_marks
                    .get(ticker)
                    .copied()
                    .unwrap_or(position.avg_entry_price);
                PortfolioPosition {
                    ticker: ticker.clone(),
                    order_type: if position.qty >= 0.0 { "long" } else { "short" }.to_owned(),
                    deposit_ratio: (position.qty.abs() * price / denominator).clamp(0.0, 1.0),
                    volume_weighted_avg_entry_price: position.avg_entry_price,
                }
            })
            .collect();
        InputDataPoint::Portfolio {
            cash: self.cash.max(0.0),
            equity,
            buying_power: self.cash.max(0.0),
            positions,
        }
    }

    pub fn traded_tickers(&self) -> BTreeSet<String> {
        self.trades
            .iter()
            .filter(|trade| trade.valid)
            .map(|trade| trade.ticker.clone())
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn order(direction: &str, ratio: f64) -> OutputDataPoint {
        OutputDataPoint::MarketOrder {
            ticker: "SPY".to_owned(),
            direction: direction.to_owned(),
            deposit_ratio: ratio,
            short_explanation: "test".to_owned(),
        }
    }

    #[test]
    fn buy_then_sell_round_trip_matches_cash_accounting() {
        let mut portfolio = Portfolio::new(10_000.0, "SPY".to_owned(), 1.0);
        let prices = BTreeMap::from([("SPY".to_owned(), 100.0)]);
        portfolio.apply_outputs(&[order("buy", 1.0)], &prices, 1, true);
        assert_eq!(portfolio.cash, 0.0);
        assert_eq!(portfolio.positions["SPY"].qty, 100.0);

        let exit_prices = BTreeMap::from([("SPY".to_owned(), 110.0)]);
        portfolio.apply_outputs(&[order("sell", 1.0)], &exit_prices, 2, true);
        assert_eq!(portfolio.cash, 11_000.0);
        assert!(!portfolio.positions.contains_key("SPY"));
        assert_eq!(portfolio.trades.len(), 2);
    }

    #[test]
    fn buy_batch_over_available_cash_records_invalid_orders() {
        let mut portfolio = Portfolio::new(10_000.0, "SPY".to_owned(), 1.0);
        let prices = BTreeMap::from([("SPY".to_owned(), 100.0)]);
        portfolio.apply_outputs(&[order("buy", 0.75), order("buy", 0.75)], &prices, 1, true);
        assert_eq!(portfolio.cash, 10_000.0);
        assert!(portfolio.positions.is_empty());
        assert_eq!(portfolio.trades.len(), 2);
        assert!(portfolio.trades.iter().all(|trade| !trade.valid));
    }
}
